libp2p/protocols/pubsub/gossipsub.nim

```diff
@@ -455,48 +463,48 @@ method unsubscribe*(g: GossipSub,
 
 method publish*(g: GossipSub,
                 topic: string,
-                data: seq[byte]) {.async.} =
-  await procCall PubSub(g).publish(topic, data)
+                data: seq[byte]): Future[int] {.async.} =
+  # base returns always 0
+  discard await procCall PubSub(g).publish(topic, data)
   trace "about to publish message on topic", name = topic,
                                              data = data.shortLog
 
   var peers: HashSet[string]
-  # TODO: we probably don't need to try multiple times
-  if data.len > 0 and topic.len > 0:
-    for _ in 0..<5: # try to get peers up to 5 times
-      if peers.len > 0:
-        break
-
-      if topic in g.topics: # if we're subscribed to the topic attempt to build a mesh
-        await g.rebalanceMesh(topic)
-        peers = g.mesh.getOrDefault(topic)
-      else: # send to fanout peers
-        await g.replenishFanout(topic)
-        if topic in g.fanout:
-          peers = g.fanout.getOrDefault(topic)
-          # set the fanout expiry time
-          g.lastFanoutPubSub[topic] = Moment.fromNow(GossipSubFanoutTTL)
 
-      # wait a second between tries
-      await sleepAsync(1.seconds)
+  if topic.len > 0: # data could be 0/empty
+    if topic in g.topics: # if we're subscribed use the mesh
+      peers = g.mesh.getOrDefault(topic)
+    else: # not subscribed, send to fanout peers
+      peers = g.fanout.getOrDefault(topic)
```

Comment:
you wont have any fanout peers here, those need to be sourced first

Reply:
I will put back the sourcing here.. but the gossipsub table being drained every heartbeat issue persists

Reply:
Just checked, this is in-line with go implementation indeed , will restore it and try go figure the heartbeat issue

Reply:
but in go they also do some extra maintenance and replenish in the heartbeat too 
https://github.com/libp2p/go-libp2p-pubsub/blob/8945f91465bc3ba418688684a2c29c384cd79647/gossipsub.go#L1436
not a big deal afaik tho

Reply:
thats a good catch, we should be replenishing peers for fanout topics that we already know about, so a `replenishFanout()` should follow the `dropFanoutPeers`, see https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.0.md#fanout-maintenance.

Reply:
I added those bits